### DISCUSSION TITLE
UseBoundStore / WithReact TS error

### DISCUSSION THREAD

**Discussion by digitlninja** [pmndrs/zustand#1807] [Q&A]
Summary
I'm getting this TS error when attempting to type our store 
"Type 'StoreApi<ShopState>' does not satisfy the constraint 'WithReact<StoreApi<object>>'.
  Type 'StoreApi<ShopState>' is not assignable to type 'StoreApi<object>'.
    Type '{}' is missing the following properties from type 'ShopState': order, setOrder"

```js
import create, { StoreApi, UseBoundStore } from 'zustand';
import { persist } from 'zustand/middleware';
import { immer } from 'zustand/middleware/immer';

interface ShopState {
  order: number;
  setOrder: (order: number) => void;
}

export interface Shop {
  id: number;
  currencyPair: string;
}

const shops: { [keyof: number]: UseBoundStore<StoreApi<ShopState>> } = {};
const useShopStore = (shop: Shop ) => {
  if (!shops[shop.id]) {
    const shopState= create<ShopState>()(
      immer((set, get) => ({
        order: 0, // set default value for order
        setOrder: (order: number) => {
          set((state) => {
            state.order = order;
          });
        },
      }))
    );
    shops[shop.id] = shopState;
  }
};
```
This very much seems like a typing issue, or an issue with versioning. I just can't figure it out. 

## Repro
https://codesandbox.io/s/zen-sound-qjemzy?file=/src/App.tsx

## Check List

Please do not ask questions in issues.

- [x ] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x ] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [x ] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment]
Here's the TS playground repro: https://tsplay.dev/mblnBw

So far, it doesn't error. Can you modify it to repro?

**Comment by digitlninja** [comment]
@dai-shi  it errors here  https://codesandbox.io/s/zen-sound-qjemzy?file=/src/App.tsx, and in vscode

**Reply by dai-shi** [reply]
https://codesandbox.io/s/busy-kowalevski-r2oh14?file=/src/App.tsx using the latest version instead of `4.0.0-rc.1`.

**Comment by digitlninja** [comment]
![image](https://github.com/pmndrs/zustand/assets/12513193/0d68b464-aec9-4471-8f8a-4caee71a4ab6)

**Reply by dbritto-dev** [reply]
@digitlninja I'll create you a codesandbox example to implement that

**Reply by dbritto-dev** [reply]
@digitlninja here you go https://codesandbox.io/s/bold-star-yzfon2?file=/src/App.tsx

We can do something like this to get the right types: 

```ts
type ImmerStateCreator<T> = StateCreator<
  T,
  [["zustand/immer", never], never],
  [],
  T
>;

type ImmerStoreApi<T> = Parameters<ImmerStateCreator<T>>[2];
```

and then we can use it like this:

```ts
const shops: Record<number, UseBoundStore<ImmerStoreApi<ShopState>>> = {};
```

> **Note:** we need `UseBoundStore<WithImmer<StoreApi<ShopState>>>` instead of `UseBoundStore<StoreApi<ShopState>>`

**Reply by digitlninja** [reply]
@dbritto-dev  Woah, thats quite a lot of wordy code. 
Im getting this new ts error now?
![image](https://github.com/pmndrs/zustand/assets/12513193/03e4c7ec-bc38-4006-9810-379051b683c2)

Aside, I've also been looking for an explanation in the docs and repo of what exactly a 'bounded' store is meant to be and its usage?
is it just a Data structure combination of multiple stores?

Thanks for your help man! Really appreciate you :)
Also, is there a discord?

**Reply by dbritto-dev** [reply]
> @dbritto-dev  Woah, thats quite a lot of wordy code. 
> Im getting this new ts error now?
> ![image](https://github.com/pmndrs/zustand/assets/12513193/03e4c7ec-bc38-4006-9810-379051b683c2)
> 
> Aside, I've also been looking for an explanation in the docs and repo of what exactly a 'bounded' store is meant to be and its usage?
> is it just a Data structure combination of multiple stores?
> 
> Thanks for your help man! Really appreciate you :)
> Also, is there a discord?
> 

Interesting I didn't get any errors

Which versions are you using? (react, zustand, etc)

**Reply by digitlninja** [reply]
I inherited this project, not sure we can upgrade as of yet. focus isn't on time upgrading packages and migrations needed.
zustand - 4.0.0-rc.1 | react - 18.1.0
please advise soonest :)

**Reply by digitlninja** [reply]
new error
![image](https://github.com/pmndrs/zustand/assets/12513193/c6b5f2ff-6111-4d7e-9474-ca5b4af933ce)
**here** => https://codesandbox.io/s/zen-sound-qjemzy?file=/src/App.tsx

how can I fix this?

**Reply by dbritto-dev** [reply]
@digitlninja would you mind sharing your new code? I'm asking for that because I was able to see double WithImmer type in your example. BTW, I encourage you to use zustand v4.3.x instead of 4.0.0-rc there're lots of fixes including type fixes.

**Reply by digitlninja** [reply]
> @digitlninja would you mind sharing your new code? I'm asking for that because I was able to see double WithImmer type in your example. BTW, I encourage you to use zustand v4.3.x instead of 4.0.0-rc there're lots of fixes including type fixes.

Thanks! What is the migration/upgrade process and ease like though?

Where is double WithImmer?

**Reply by dbritto-dev** [reply]
> > @digitlninja would you mind sharing your new code? I'm asking for that because I was able to see double WithImmer type in your example. BTW, I encourage you to use zustand v4.3.x instead of 4.0.0-rc there're lots of fixes including type fixes.
> 
> Thanks! What is the migration/upgrade process and ease like though?
> 
> Where is double WithImmer?

@digitlninja yeah, shouldn't be an issue since you are using v4.0.0-rc
